src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -123,58 +119,58 @@ private unsafe void ParseRequestLine(TRequestHandler handler, byte* data, int le
                 }
             }
 
-            var pathBuffer = new Span<byte>(data + pathStart, offset - pathStart);
+            var path = new TargetOffsetPathLength(targetStart, length: offset - targetStart, pathEncoded);
 
             // Query string
-            var queryStart = offset;
             if (ch == ByteQuestionMark)
             {
                 // We have a query string
-                for (; offset < length; offset++)
+                for (; (uint)offset < (uint)requestLine.Length; offset++)
                 {
-                    ch = data[offset];
+                    ch = requestLine[offset];
                     if (ch == ByteSpace)
                     {
                         break;
                     }
                 }
             }
 
-            // End of query string not found
-            if (offset == length)
-            {
-                RejectRequestLine(data, length);
-            }
-
-            var targetBuffer = new Span<byte>(data + pathStart, offset - pathStart);
-            var query = new Span<byte>(data + queryStart, offset - queryStart);
-
+            var queryEnd = offset;
             // Consume space
             offset++;
 
+            // If offset has overshot length, end of query string wasn't not found
+            if ((uint)offset > (uint)requestLine.Length)
+            {
+                goto Reject;
```

Comment:
😢 gotos?

Reply:
In place calls to `RejectRequestLine` adds 3 extra chunks of
```asm
G_M45614_IG42:
       mov      rcx, rsi
       lea      rdx, bword ptr [rsp+40H]
       mov      bword ptr [rdx], rbx
       mov      dword ptr [rdx+8], ebp
       lea      rdx, bword ptr [rsp+40H]
       call     HttpParser:RejectRequestLine(System.ReadOnlySpan`1[Byte]):this
       int3     
```
Changing method's asm from 
```asm
; Total bytes of code 892
```
to
```asm
; Total bytes of code 967
```

Reply:
https://github.com/dotnet/aspnetcore/pull/20562 worked with the calls as @AndyAyersMS added a PR to runtime to merge throwhelper calls to a single call (can't find the PR)

However its not working here (hence the gotos); this is perhaps due to the `if` in the throwhelper https://github.com/dotnet/aspnetcore/pull/20885/commits/56c03726f05ee39f824c1e3b708e0c5beb9f0c45#diff-d6a9ae754775b6cb72445c3b8db610c2R499-R503
```csharp
throw GetInvalidRequestException(
	IsTlsHandshake(requestLine) ?
	RequestRejectionReason.TlsOverHttpError :
	RequestRejectionReason.InvalidRequestLine,
	requestLine);
```
I'll move it behind a call and see if that allows the merge so can use direct calls rather than gotos

Reply:
Alas still ends up with 4 calls. It does move them to the end of the method (for current and extra indirection) so it does recognise them as throwhelpers; just doesn't merge them...

Reply:
It was dotnet/coreclr#27113.

If you drop me a copy of the assembly(s) I can take a look.

Reply:
> If you drop me a copy of the assembly(s) I can take a look.

Compilable extract https://gist.github.com/benaadams/c81babbac40969a002e7fda0f67dead4

`HttpParser.ParseRequestLine` is the method; its hooked up in `Main`

Reply:
The throw helper calls are `RejectRequestLine(requestLine)`

Reply:
We are currently running the transformation after morph, and morph is introducing different temps for the different calls, so the call trees aren't identical.

I think I can move this transformation earlier and run it before morph; I wasn't able to do this initially because we lacked the right control flow information, but I've since fixed that.

Before morph
```
[000225] --C-G-------              *  CALL      void   HttpParser.RejectRequestLine
[000223] ------------ this in rcx  +--*  LCL_VAR   ref    V00 this         
[000227] n----------- arg1         \--*  OBJ       struct<ReadOnlySpan`1, 16>
[000226] ------------                 \--*  ADDR      byref 
[000224] ------------                    \--*  LCL_VAR   struct<ReadOnlySpan`1, 16> V02 arg2         

[000174] --C-G-------              *  CALL      void   HttpParser.RejectRequestLine
[000172] ------------ this in rcx  +--*  LCL_VAR   ref    V00 this         
[000176] n----------- arg1         \--*  OBJ       struct<ReadOnlySpan`1, 16>
[000175] ------------                 \--*  ADDR      byref 
[000173] ------------                    \--*  LCL_VAR   struct<ReadOnlySpan`1, 16> V02 arg2         
```
After morph
```
[000225] -ACXG+------              *  CALL      void   HttpParser.RejectRequestLine
[001106] -A--------L- this SETUP   +--*  ASG       ref   
[001105] D------N----              |  +--*  LCL_VAR   ref    V104 tmp90       
[000223] -----+------              |  \--*  LCL_VAR   ref    V00 this         
[001104] -A---+----L- arg1 SETUP   +--*  COMMA     void  
[001097] -A----------              |  +--*  COMMA     void  
[001092] -A----------              |  |  +--*  ASG       byref 
[001091] D------N----              |  |  |  +--*  LCL_VAR   byref  V103 tmp89       
[001089] -----+------              |  |  |  \--*  ADDR      byref 
[001090] ----G+-N----              |  |  |     \--*  LCL_VAR   struct<ReadOnlySpan`1, 16>(AX) V96 tmp82        
[001096] -A----------              |  |  \--*  ASG       byref 
[001094] *------N----              |  |     +--*  IND       byref 
[001093] ------------              |  |     |  \--*  LCL_VAR   byref  V103 tmp89        Zero Fseq[_pointer]
[001095] -------N----              |  |     \--*  LCL_VAR   byref  V56 tmp42        
[001103] -A----------              |  \--*  ASG       int   
[001101] *------N----              |     +--*  IND       int   
[001100] ------------              |     |  \--*  ADD       byref 
[001098] ------------              |     |     +--*  LCL_VAR   byref  V103 tmp89       
[001099] ------------              |     |     \--*  CNS_INT   long   8 Fseq[_length]
[001102] -------N----              |     \--*  LCL_VAR   int    V57 tmp43        
[001107] ------------ this in rcx  +--*  LCL_VAR   ref    V104 tmp90       
[001109] ------------ arg1 in rdx  \--*  ADDR      byref 
[001108] -------N----                 \--*  LCL_VAR   struct<ReadOnlySpan`1, 16>(AX) V96 tmp82        

[000174] -ACXG+------              *  CALL      void   HttpParser.RejectRequestLine
[001206] -A--------L- this SETUP   +--*  ASG       ref   
[001205] D------N----              |  +--*  LCL_VAR   ref    V109 tmp95       
[000172] -----+------              |  \--*  LCL_VAR   ref    V00 this         
[001204] -A---+----L- arg1 SETUP   +--*  COMMA     void  
[001197] -A----------              |  +--*  COMMA     void  
[001192] -A----------              |  |  +--*  ASG       byref 
[001191] D------N----              |  |  |  +--*  LCL_VAR   byref  V108 tmp94       
[001189] -----+------              |  |  |  \--*  ADDR      byref 
[001190] ----G+-N----              |  |  |     \--*  LCL_VAR   struct<ReadOnlySpan`1, 16>(AX) V96 tmp82        
[001196] -A----------              |  |  \--*  ASG       byref 
[001194] *------N----              |  |     +--*  IND       byref 
[001193] ------------              |  |     |  \--*  LCL_VAR   byref  V108 tmp94        Zero Fseq[_pointer]
[001195] -------N----              |  |     \--*  LCL_VAR   byref  V56 tmp42        
[001203] -A----------              |  \--*  ASG       int   
[001201] *------N----              |     +--*  IND       int   
[001200] ------------              |     |  \--*  ADD       byref 
[001198] ------------              |     |     +--*  LCL_VAR   byref  V108 tmp94       
[001199] ------------              |     |     \--*  CNS_INT   long   8 Fseq[_length]
[001202] -------N----              |     \--*  LCL_VAR   int    V57 tmp43        
[001207] ------------ this in rcx  +--*  LCL_VAR   ref    V109 tmp95       
[001209] ------------ arg1 in rdx  \--*  ADDR      byref 
[001208] -------N----                 \--*  LCL_VAR   struct<ReadOnlySpan`1, 16>(AX) V96 tmp82        
```

Reply:
Resolved by https://github.com/dotnet/runtime/pull/35255; thank you @AndyAyersMS 😀